File: articles/ai-native-ways/09-embedded/example-1/sensor_logic.py
#!/usr/bin/env python3
"""温室の温度制御ロジック ── まず Python で書いて、シミュレーションで詰める。

仕様:
- 温度センサが 1 分ごとに値を返す(℃, 浮動小数点)
- 設定: 目標温度 22℃、ヒステリシス ±1.5℃
- ヒーターを ON/OFF する
- 連続稼働時間が 30 分を超えたら必ず 5 分休む(過熱防止)
- 5 分間センサ無応答ならフェイルセーフでヒーター OFF + 警報

これを実機(ESP32 + MicroPython)に移植する前に、ここで完全に検証する。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class HeaterController:
    target: float = 22.0
    hysteresis: float = 1.5
    max_on_minutes: int = 30
    forced_rest_minutes: int = 5
    sensor_timeout_minutes: int = 5

    # state
    heater_on: bool = False
    on_run_minutes: int = 0
    forced_rest_remaining: int = 0
    silent_minutes: int = 0
    alarm: bool = False

    # log
    events: list[str] = field(default_factory=list)

    def tick(self, t_minute: int, sensor_temp: Optional[float]) -> None:
        # フェイルセーフ: センサ無応答
        if sensor_temp is None:
            self.silent_minutes += 1
            if self.silent_minutes >= self.sensor_timeout_minutes:
                if self.heater_on or not self.alarm:
                    self.events.append(f"t={t_minute:>4} ALARM ヒーター強制 OFF (センサ {self.silent_minutes} 分無応答)")
                self.heater_on = False
                self.alarm = True
            return
        else:
            if self.silent_minutes >= self.sensor_timeout_minutes:
                self.events.append(f"t={t_minute:>4} 復旧 (センサ復活)")
            self.silent_minutes = 0
            self.alarm = False

        # 強制休止中
        if self.forced_rest_remaining > 0:
            self.forced_rest_remaining -= 1
            if self.heater_on:
                self.heater_on = False
                self.events.append(f"t={t_minute:>4} 強制 OFF (連続 {self.max_on_minutes} 分到達)")
            return

        # 通常制御
        if self.heater_on:
            self.on_run_minutes += 1
            if self.on_run_minutes >= self.max_on_minutes:
                self.heater_on = False
                self.forced_rest_remaining = self.forced_rest_minutes
                self.on_run_minutes = 0
                self.events.append(f"t={t_minute:>4} 強制 OFF + {self.forced_rest_minutes} 分休止開始")
                return
            if sensor_temp >= self.target + self.hysteresis:
                self.heater_on = False
                self.on_run_minutes = 0
                self.events.append(f"t={t_minute:>4} OFF ({sensor_temp:.1f}℃ ≥ {self.target + self.hysteresis})")
        else:
            if sensor_temp <= self.target - self.hysteresis:
                self.heater_on = True
                self.on_run_minutes = 0
                self.events.append(f"t={t_minute:>4} ON  ({sensor_temp:.1f}℃ ≤ {self.target - self.hysteresis})")

File: articles/examples/ai-native-ways/08-embedded/example-1/sensor_logic.c
/*
 * 同じロジックの C 実装(ESP32 等の実機への移植イメージ)
 *
 * Python の sensor_logic.py を直接翻訳したもの。Claude に
 *   「これを ESP-IDF で動く C に翻訳して」と頼むだけで出てくる。
 * Python 側で完全に検証してからこちらに移すので、ロジックの
 * バグは混入しない。あとはハードウェア依存のピン制御だけ。
 */

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <math.h>

// ハードウェア層 (実機では GPIO・I2C・タイマ駆動)
extern float read_temperature_sensor(void);   // -INFINITY なら無応答
extern void  set_heater(bool on);
extern void  trigger_alarm(const char *msg);
extern void  log_event(const char *msg);

#define TARGET                  22.0f
#define HYSTERESIS               1.5f
#define MAX_ON_MINUTES          30
#define FORCED_REST_MINUTES      5
#define SENSOR_TIMEOUT_MINUTES   5

typedef struct {
    bool     heater_on;
    uint8_t  on_run_minutes;
    uint8_t  forced_rest_remaining;
    uint8_t  silent_minutes;
    bool     alarm;
} controller_t;

static controller_t ctrl;

void controller_tick(uint32_t minute) {
    float temp = read_temperature_sensor();
    bool  sensor_ok = !isinf(temp);

    if (!sensor_ok) {
        ctrl.silent_minutes++;
        if (ctrl.silent_minutes >= SENSOR_TIMEOUT_MINUTES) {
            if (ctrl.heater_on || !ctrl.alarm) {
                char buf[80];
                snprintf(buf, sizeof buf,
                         "t=%4u ALARM ヒーター強制 OFF (センサ %u 分無応答)",
                         minute, ctrl.silent_minutes);
                log_event(buf);
                trigger_alarm(buf);
            }
            ctrl.heater_on = false;
            set_heater(false);
            ctrl.alarm = true;
        }
        return;
    }

    if (ctrl.silent_minutes >= SENSOR_TIMEOUT_MINUTES) {
        char buf[80];
        snprintf(buf, sizeof buf, "t=%4u 復旧 (センサ復活)", minute);
        log_event(buf);
    }
    ctrl.silent_minutes = 0;
    ctrl.alarm = false;

    if (ctrl.forced_rest_remaining > 0) {
        ctrl.forced_rest_remaining--;
        if (ctrl.heater_on) {
            ctrl.heater_on = false;
            set_heater(false);
        }
        return;
    }

    if (ctrl.heater_on) {
        ctrl.on_run_minutes++;
        if (ctrl.on_run_minutes >= MAX_ON_MINUTES) {
            ctrl.heater_on = false;
            set_heater(false);
            ctrl.forced_rest_remaining = FORCED_REST_MINUTES;
            ctrl.on_run_minutes = 0;
            return;
        }
        if (temp >= TARGET + HYSTERESIS) {
            ctrl.heater_on = false;
            set_heater(false);
            ctrl.on_run_minutes = 0;
        }
    } else {
        if (temp <= TARGET - HYSTERESIS) {
            ctrl.heater_on = true;
            set_heater(true);
            ctrl.on_run_minutes = 0;
        }
    }
}
